File: heap-sort/python/heap-sort.py
#!/usr/bin/python
#coding:utf8

def swap(list, i, j):
    temp = list[i]
    list[i] = list[j]
    list[j] = temp

def findLeft(i):
    return 2*i+1

def findRight(i):
    return 2*i+2

def keepMaxHeap(list, i, heapSize):
    left = findLeft(i)
    right = findRight(i)
    max = i
    if left <= heapSize and list[left] > list[max]:
	max = left
    if right <= heapSize and list[right] > list[max]:
	max = right
    if max != i:
	swap(list, i, max)
	keepMaxHeap(list, max, heapSize)

def heapSort(list):
    length = len(list)
    for i in range(length/2, -1, -1):
        keepMaxHeap(list, i, length-1)
    for i in range(length-1, 0, -1):
	swap(list, 0, i)
	keepMaxHeap(list, 0, i-1)

list = [4, 2, 8, 5, 7, 0, 3, 9, 1, 6]

print 'before heap sort: ', list

heapSort(list)

print 'after heap sort: ', list

File: heap-sort/java/HeapSort.java
public class HeapSort {
	public static void main(String[] args) {
		int[] data = {3, 2, 9, 5, 7, 1, 4, 6, 0, 8};
		System.out.print("before heap Sort: ");
		printArray(data);
		System.out.println();
		heapSort(data);
		System.out.print("after heap Sort: ");
		printArray(data);
	}

	public static void heapSort(int[] data) {
		for (int i = data.length / 2; i >= 0; i--) {
			keepMaxHeap(data, i, data.length - 1);
		}

		for (int i = data.length - 1; i >= 0; i--) {
			swap(data, i, 0);
			keepMaxHeap(data, 0, i - 1);
		}
	}

	public static void keepMaxHeap(int[] data, int i, int heapSize) {
		int left = 2 * i + 1;
		int right = 2 * i + 2;
		int max = i;
		if(left <= heapSize && data[left] > data[max]) {
			max = left;
		}
		if(right <= heapSize && data[right] > data[max]) {
			max = right;
		}
		if(max != i) {
			swap(data, i, max);
			keepMaxHeap(data, max, heapSize);
		}
	}

	public static void swap(int[] data, int p, int q) {
		int temp = data[p];
		data[p] = data[q];
		data[q] = temp;
	}

	public static void printArray(int[] data) {
		for(int i : data) {
			System.out.print(i + " ");
		}
	}
}